### PR TITLE
Add support for global proxy in openshift

### PR DESCRIPTION
**Pull request by jkhelil**

### CODE REVIEW

File: pkg/configuration/base/resources/proxy.go [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+		if found {
+			overrides = true
+			return
+		}
+	}
+
+	return
+}
+
+func find(proxyEnvVar []corev1.EnvVar, name string) (envVar *corev1.EnvVar, found bool) {
```
Comment (waveywaves) [inline pkg/configuration/base/resources/proxy.go]: please rename function to something more specific like getProxyEnvVar or findProxyEnvVar

File: pkg/configuration/base/resources/proxy.go [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+	ProxyAPIChecked     = false
+	allProxyEnvVarNames = []string{
+		envHTTPProxyName,
+		envHTTPSProxyName,
+		envNoProxyName,
+	}
+	ProxyEnvVars = []corev1.EnvVar{}
+)
+
+func IsProxyAPIAvailable(clientSet *kubernetes.Clientset) bool {
```
Comment (waveywaves) [inline pkg/configuration/base/resources/proxy.go]: rename to IsOpenshiftProxyAPIAvailable

File: pkg/configuration/base/resources/proxy.go [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+		ProxyAPIAvailable = true
+	}
+	return ProxyAPIAvailable
+}
+
+func QueryProxyConfig(k8sClient client.Client) error {
+	global := &apiconfigv1.Proxy{}
+	err := k8sClient.Get(context.TODO(), types.NamespacedName{Name: DefaultOpenShiftProxyNamespace}, global)
+	if err != nil {
+		logger.Info(fmt.Sprintf("global proxy not found in namespace %s", DefaultOpenShiftProxyNamespace))
```
Comment (waveywaves) [inline pkg/configuration/base/resources/proxy.go]: "openshift global proxy not found in namespace %s" would be more specific

**Review by waveywaves — CHANGES_REQUESTED** [review]
Considering this is openshift specific, we should reflect that in the functions and variable names as well. Otherwise it could be misleading regarding the proxy support.

File: pkg/configuration/base/resources/proxy.go (hunk 1 above)
Comment (jkhelil) [reply pkg/configuration/base/resources/proxy.go]: done

File: pkg/configuration/base/resources/proxy.go (hunk 3 above)
Comment (jkhelil) [reply pkg/configuration/base/resources/proxy.go]: done

File: pkg/configuration/base/resources/proxy.go (hunk 2 above)
Comment (jkhelil) [reply pkg/configuration/base/resources/proxy.go]: done

File: pkg/configuration/base/resources/pod.go [hunk 4]
```diff
@@ -232,6 +232,11 @@ func NewJenkinsMasterContainer(jenkins *v1alpha2.Jenkins) corev1.Container {
 		envs = append(envs, jenkinsHomeEnvVar)
 	}
 
+	if !IsProxyOverridden(envs) {
+		ProxyEnvVars = dropEmptyProxyEnv(ProxyEnvVars)
+		envs = append(envs, ProxyEnvVars...)
```
Comment (akram) [inline pkg/configuration/base/resources/pod.go:237]: shouldn't you remove ProxyEnvVars by key instead ?

File: pkg/configuration/base/resources/proxy.go [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+	if ProxyAPIChecked {
+		return ProxyAPIAvailable
+	}
+	gv := schema.GroupVersion{
+		Group:   apiconfigv1.GroupName,
+		Version: apiconfigv1.SchemeGroupVersion.Version,
+	}
+	if err := discovery.ServerSupportsVersion(clientSet, gv); err != nil {
+		// error, API not available
+		ProxyAPIChecked = true
```
Comment (akram) [inline pkg/configuration/base/resources/proxy.go:43]: In all cases, `ProxyAPIChecked` is set to true, so, we can confidently remove it from `if/else`

File: pkg/configuration/base/resources/proxy.go [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+		ProxyAPIAvailable = false
```
Comment (akram) [inline pkg/configuration/base/resources/proxy.go:44]: You can also set it to false before the if/else, and remove the else.

File: pkg/configuration/base/resources/proxy.go [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+
+	// We have found the global proxy configuration object!
+	ProxyEnvVars = ToEnvVar(global)
+	return nil
+}
+
+func ToEnvVar(proxy *apiconfigv1.Proxy) []corev1.EnvVar {
+	return []corev1.EnvVar{
+		{
+			Name:  envHTTPProxyName,
```
Comment (akram) [inline pkg/configuration/base/resources/proxy.go:69]: As this is a constant, can you rename it to uppercase only.
```
const (
 HTTP_PROXY = "HTTP_PROXY" 
)
```
would be ok.

File: pkg/configuration/base/resources/proxy.go [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+
+			break
+		}
+	}
+
+	return
+}
+
+func dropEmptyProxyEnv(in []corev1.EnvVar) (out []corev1.EnvVar) {
+	out = make([]corev1.EnvVar, 0)
```
Comment (akram) [inline pkg/configuration/base/resources/proxy.go:114]: can you rename `out` to a more explanatory name?

**Review by akram — CHANGES_REQUESTED** [review]

File: pkg/configuration/base/resources/pod.go (hunk 4 above)
Comment (jkhelil) [reply pkg/configuration/base/resources/pod.go:237]: No it is good like this

File: pkg/configuration/base/resources/proxy.go (hunk 5 above)
Comment (jkhelil) [reply pkg/configuration/base/resources/proxy.go:43]: done

File: pkg/configuration/base/resources/proxy.go (hunk 6 above)
Comment (jkhelil) [reply pkg/configuration/base/resources/proxy.go:44]: done

File: pkg/configuration/base/resources/proxy.go (hunk 7 above)
Comment (jkhelil) [reply pkg/configuration/base/resources/proxy.go:69]: done

File: pkg/configuration/base/resources/proxy.go (hunk 8 above)
Comment (jkhelil) [reply pkg/configuration/base/resources/proxy.go:114]: it is just a helper method, it is clear like this

File: pkg/configuration/base/resources/proxy.go (hunk 7 above)
Comment (jkhelil) [reply pkg/configuration/base/resources/proxy.go:69]: we should Camelcase and HTTPPROXY is ugly, lets keep it

**Pull request merged**